lib/ansible/modules/storage/glusterfs/gluster_heal_facts.py

```diff
@@ -0,0 +1,191 @@
+    required: true
+    aliases: ['volume']
+  status_filter:
+    default: "self-heal"
+    choices: ["self-heal", "rebalance"]
+    description:
+      - Determines which facts are to be returned.
+      - If the C(status_filter) is C(self-heal), status of self-heal, along with the number of files still in process are returned.
+      - If the C(status_filter) is C(rebalance), rebalance status is returned.
+        self-heal, status of self-heal, along with the number of files still
+        in process are returned. If status_filter is rebalance,
+        rebalance status is returned.
+requirements:
+  - GlusterFS > 3.2
+'''
+
+EXAMPLES = '''
+- name: Gather self-heal facts about all gluster hosts in the cluster
+  gluster_heal_facts:
+    name: test_volume
+    status_filter: self-heal
+  register: self_heal_status
+- debug:
+    var: self_heal_status
+# status along with the number of files still in process:
+- gluster_heal_facts:
+    name: test_volume
+    status_filter: self-heal
+- debug:
+    var: gluster_hosts
+'''
+
+RETURN = '''
+name:
+    description: GlusterFS volume name
+    returned: always
+    type: str
+status_filter:
+    description: Whether self-heal or rebalance status is to be returned
+    returned: always
+    type: str
+heal_info:
+    description: List of files that still need healing process
+    returned: On success
+    type: list
+rebalance_status:
+    description: Status of rebalance operation
+    returned: On success
+    type: list
+'''
+
+import traceback
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils._text import to_native
+from distutils.version import LooseVersion
+
+glusterbin = ''
+
+
+def run_gluster(gargs, **kwargs):
+    global glusterbin
+    global module
+    args = [glusterbin, '--mode=script']
+    args.extend(gargs)
+    try:
+        rc, out, err = module.run_command(args, **kwargs)
+        if rc != 0:
+            module.fail_json(msg='error running gluster (%s) command (rc=%d): %s' %
+                             (' '.join(args), rc, out or err), exception=traceback.format_exc())
+    except Exception as e:
+        module.fail_json(msg='error running gluster (%s) command: %s' % (' '.join(args),
+                                                                         to_native(e)), exception=traceback.format_exc())
+    return out
+
+
+def get_self_heal_status(name):
+    out = run_gluster(['volume', 'heal', name, 'info'], environ_update=dict(LANG='C', LC_ALL='C', LC_MESSAGES='C'))
+    raw_out = out.split("\n")
+    heal_info = []
+    # return files that still need healing.
+    for line in raw_out:
+        if 'Brick' in line:
+            br_dict = {}
+            br_dict['brick'] = line.strip().strip("Brick")
+        elif 'Status' in line:
+            br_dict['status'] = line.split(":")[1].strip()
+        elif 'Number' in line:
+            br_dict['no_of_entries'] = line.split(":")[1].strip()
+            if int(br_dict['no_of_entries']) != 0:
```

Comment:
This will throw an exception every time when a server is down. When glusterfsd is down the output looks like this:

Brick 10.70.43.200:/mnt/engine
Status: Transport endpoint is not connected
Number of entries: -

And you'll be trying to do int('-') which will throw ValueError. And the module throws error:
fatal: [10.70.42.25]: FAILED! => {"changed": false, "msg": "Invalid heal status option."}                                       
in the function main.

Reply:
done.